aws/sdk/integration-tests/ec2/tests/waiters.rs

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use aws_sdk_ec2::{client::Waiters, config::Region, error::DisplayErrorContext, Client};
+use aws_smithy_async::test_util::tick_advance_sleep::{
+    tick_advance_time_and_sleep, TickAdvanceTime,
+};
+use aws_smithy_runtime::{
+    client::http::test_util::dvr::ReplayingClient, test_util::capture_test_logs::show_test_logs,
+};
+use aws_smithy_runtime_api::client::waiters::error::WaiterError;
+use aws_smithy_types::retry::RetryConfig;
+use std::time::Duration;
+
+async fn prerequisites() -> (Client, ReplayingClient, TickAdvanceTime) {
+    let (time_source, sleep_impl) = tick_advance_time_and_sleep();
+    let client =
+        ReplayingClient::from_file("tests/instance-status-ok-waiter-success.json").unwrap();
+    let config = aws_sdk_ec2::Config::builder()
+        .with_test_defaults()
+        .http_client(client.clone())
+        .time_source(time_source.clone())
+        .sleep_impl(sleep_impl)
+        .region(Region::new("us-west-2"))
+        .retry_config(RetryConfig::standard())
+        .build();
+    (aws_sdk_ec2::Client::from_conf(config), client, time_source)
+}
+
+#[tokio::test]
+async fn waiters_success() {
+    let _logs = show_test_logs();
+
+    let (ec2, http_client, time_source) = prerequisites().await;
+
+    ec2.start_instances()
+        .instance_ids("i-09fb4224219ac6902")
+        .send()
+        .await
+        .unwrap();
+
+    let waiter_task = tokio::spawn(
+        ec2.wait_for_instance_status_ok()
+            .instance_ids("i-09fb4224219ac6902")
+            .wait(Duration::from_secs(300)),
+    );
+
+    time_source.tick(Duration::from_secs(305)).await;
```

Comment:
**question**: Why do we need to advance this far? I would assume the waiter finishes before this in the success case and I found this slightly odd when reading the test.

Reply:
My guess is that `time_source` and `sleep_impl` returned by `tick_advance_time_and_sleep` are special, i.e. they are frozen unless `tick` is called.

Reply:
Yuki is right. The responses in the test data require that the waiter wait around and poll a few times, so time has to get advanced. However, it doesn't need to get advanced a full 300 seconds for the success case, as Aaron points out. I had it advance that far so that in the case of failure, it would still complete (max time exceeded).

Reply:
Added a comment in 637b4b6835607b6340cf3a15200662392034b12c